### DISCUSSION TITLE
Rouge referral problem in Google Analytics

### DISCUSSION THREAD

**Discussion by jorgenboganes** [vercel/next.js#50583] [Help]
### Summary

I have a next.js website that is experiencing the rouge referral problem. It seems that this problem is well known, and has been [present on single-page apps since at least 2016](https://www.simoahava.com/gtm-tips/fix-rogue-referral-problem-single-page-sites/). I have tried the fixes in the article, but am still experiencing the same errors. This problem is time critical, and I am struggling to find a solution. 

The problem occurs when the payment provider takes the user to another website, and then back to ours. This makes the website lose its state, and google then doesn't know who the referrer is. We are using the library react-ga.

Earlier we had a static multi-page site that got a new "page" in google tag assistant every time you changed pages. In the new single-page version however, there is only ever one page. I think this might be related to the problem?

I wonder if any of you have experienced this with a nextjs app and have any suggestions on how to solve the problem. I am not used to the Google Analytics-realm, so I apologize if there is not enough info here to go by. Plaese let me know if any other information is needed!

Thanks!

**Comment by icyJoseph** [comment]
Yeah one way I fixed this was (very ugly 🙈🙈🙈🙈), to:

```js

useEffect(() => {
  let [referrer, href] = [document.referrer, window.location.href];

  const handler = () => {
   [referrer, href] = [href, window.location.href];
   
   // pass information to 3rd parties
  }
  Router.events.on("routeChangeComplete", handler);

  return () => { 
  Router.events.off("routeChangeComplete", handler);
  }
}, [])
```

But hey, now in appDir there's no Router events.

**Reply by jorgenboganes** [reply]
Won't this just reset the referrer whenever the customer comes from another website? That's what I'm trying to avoid. And obviously I could be saving this in local storage or something, but that sounds like a problem that should have already been solved...

**Reply by icyJoseph** [reply]
Could you explain your problem in steps? 

Also, are you creating a page view on SPA route transitions?

If I understand correctly your problem is:

- User navigates to your site from A.com (GA should see referrer is now A.com)
- User navigates around in your site (GA sees referrer is still A.com ,unless you let GA know that it has changed)
- At some point in the user session they click something that takes them to B.com
- B.com returns them back to your site
- This is now a new page load so GA seems B.com as referrer

At which point is it not working the way you want? Or am I totally out of it and talking about something else? 

What should Next.js do otherwise? During SPA transitions the page is not unloaded,  so the referrer stays fixed. You can't set document.referrer, most know hacks to spoof it don't work in all browsers. It's up to you to keep track of it somehow

**Reply by jorgenboganes** [reply]
We have set up so that history change counts as page view on Google, yes. 
And you are close - The way I see it, the flow is like this:

	1. User comes to my site from A.com, referrer is "A.com"
	2. User does their shopping and goes to pay
	3. Payment provider redirects the user, for example to paypal.com, and then back to my site
	4. Now the referrer is "Google organic", and google doesn't know who to credit for the purchase/conversion

100% agree with the last part of your reply, I guess I just assumed that this problem would be persistent over the SPA-space and that it would be talked about more. So you think the only way for me to solve it is to keep track of the referrer myself, and then send it to google with/after the conversion?

Thank you for taking the time to reply, by the way!

**Reply by icyJoseph** [reply]
One time I worked with `utm_` tags, and we used that to attribute. So when the user is taken elsewhere, the return URI/URL contains a query with where we came from, I guess on page load you could parse that query and pass it to GA? 

No worries, I've always found this area of front end dev very interesting, and there's a lot of missing info, and vague assumptions from developers. I bet you there's a lot of people who have missed on this issue 😅
